runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validation_test.go

```diff
@@ -0,0 +1,1265 @@
+
+		require.NoError(t, err)
+	})
+
+	t.Run("change field type capability reference auth allowed multiple intersected fewer entitlements", func(t *testing.T) {
+
+		t.Parallel()
+
+		const oldCode = `
+			pub contract Test {
+				pub struct interface I {
+					pub fun bar()
+				}
+				pub struct interface J {
+					pub fun foo()
+				}
+				pub struct S:I, J {
+					pub fun foo() {}
+					pub fun bar() {}
+				}
+
+				pub var a: Capability<&{I, J}>?
+				init() {
+					self.a = nil
+				}
+			}
+	`
+
+		const newCode = `
+		access(all) contract Test {
+			access(all) entitlement E
+			access(all) entitlement F
+
+			access(all) struct interface I {
+				access(E) fun foo()
+			}
+			access(all) struct interface J {
+				access(F) fun bar()
+			}
+
+			access(all) struct S:I, J {
+				access(E) fun foo() {}
+				access(F) fun bar() {}
+			}
+
+			access(all) var a: Capability<auth(E) &{I, J}>?
+			init() {
+				self.a = nil
+			}
+		}
+	`
+
+		err := testContractUpdate(t, oldCode, newCode)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("change field type capability reference auth multiple intersected with too many entitlements", func(t *testing.T) {
+
+		t.Parallel()
+
+		const oldCode = `
+			pub contract Test {
+				pub struct interface I {
+					pub fun bar()
+				}
+				pub struct interface J {
+					pub fun foo()
+				}
+				pub struct S:I, J {
+					pub fun foo() {}
+					pub fun bar() {}
+				}
+
+				pub var a: Capability<&{I, J}>?
+				init() {
+					self.a = nil
+				}
+			}
+	`
+
+		const newCode = `
+		access(all) contract Test {
+			access(all) entitlement E
+			access(all) entitlement F
+
+			access(all) struct interface I {
+				access(E) fun foo()
+			}
+			access(all) struct interface J {}
+
+			access(all) struct S:I, J {
+				access(E) fun foo() {}
+			}
+
+			access(all) var a: Capability<auth(E, F) &{I, J}>?
+			init() {
+				self.a = nil
+			}
+		}
+	`
+
+		err := testContractUpdate(t, oldCode, newCode)
+
+		cause := getSingleContractUpdateErrorCause(t, err, "Test")
+		assertFieldAuthorizationMismatchError(t, cause, "Test", "a", "E", "E, F")
+	})
+
+}
+
+func TestContractUpgradeIntersectionFieldType(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("change field type restricted type", func(t *testing.T) {
+
+		t.Parallel()
+
+		const oldCode = `
+            pub contract Test {
+				pub resource interface I {}
+				pub resource R:I {}
+
+                pub var a: @R{I}
+                init() {
+                    self.a <- create R()
+                }
+            }
+        `
+
+		const newCode = `
+            access(all) contract Test {
+				access(all) resource interface I {}
+				access(all) resource R:I {}
+
+                access(all) var a: @{I} 
+                init() {
+                    self.a <- create R()
+                }
+            }
+        `
+
+		err := testContractUpdate(t, oldCode, newCode)
+
+		// This is not allowed because `@R{I}` is converted to `@R`, not `@{I}`
```

Comment:
I think this should be a valid change for restricted types. i.e: both
- `R{I}` -> `R`
- `R{I}` -> `{I}`

should be valid changes.